### PR TITLE
Improved error messaging for index signature parameters

### PR DESCRIPTION
**Pull request by kujon**
Fixes https://github.com/Microsoft/TypeScript/issues/18992

Implements this suggestion specifically: https://github.com/Microsoft/TypeScript/issues/18992#issuecomment-335458531

### CODE REVIEW

File: tests/baselines/reference/indexerConstraints2.errors.txt [hunk 1]
```diff
@@ -37,4 +44,61 @@ tests/cases/compiler/indexerConstraints2.ts(26,5): error TS2413: Numeric index t
...
+         ~
+!!! error TS1023: An index signature parameter type must be 'string' or 'number'.
+    }
+    
+    type IndexableUnion = "foo" | "bar";
+    
+    interface O {
+        [u: IndexableUnion]: A;
+         ~
+!!! error TS1337: An index signature parameter type cannot be a union type. Use '[K in IndexableUnion]' instead.
```
Comment (RyanCavanaugh) [inline tests/baselines/reference/indexerConstraints2.errors.txt]: This version of the error message should only appear when the parent is a type alias

**Comment by RyanCavanaugh** [comment]
Looks good overall, just one comment

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -25775,6 +25775,16 @@ namespace ts {
                 return grammarErrorOnNode(parameter.name, Diagnostics.An_index_signature_parameter_must_have_a_type_annotation);
             }
             if (parameter.type.kind !== SyntaxKind.StringKeyword && parameter.type.kind !== SyntaxKind.NumberKeyword) {
+                const type = getTypeFromTypeNode(parameter.type);
+
+                if (type.flags & TypeFlags.String || type.flags & TypeFlags.Number) {
+                    return grammarErrorOnNode(parameter.name, Diagnostics.An_index_signature_parameter_type_cannot_be_a_type_alias_Use_index_Colon_0_instead, typeToString(type));
+                }
+
+                if (type.flags & TypeFlags.Union && every((<UnionType>type).types, t => !!(t.flags & TypeFlags.StringLiteral))) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: `allTypesAssignableToKind`

File: src/compiler/diagnosticMessages.json [hunk 3]
```diff
@@ -939,6 +939,14 @@
         "category": "Error",
         "code": 1335
     },
+    "An index signature parameter type cannot be a type alias. Use '[index: {0}]' instead.": {
+        "category": "Error",
+        "code": 1336
+    },
+    "An index signature parameter type cannot be a union type. Use '[K in {0}]' instead.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: Nit: this won't work if there are any other members in the object type. Maybe we should just write "Consider using a mapped object type instead.

File: src/compiler/diagnosticMessages.json [hunk 4]
```diff
@@ -939,6 +939,14 @@
         "category": "Error",
         "code": 1335
     },
+    "An index signature parameter type cannot be a type alias. Use '[index: {0}]' instead.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: `Consider writing '[{0}: {1}]: {2}' instead.`

File: src/compiler/diagnosticMessages.json (hunk 3 above)
Comment (kujon) [reply src/compiler/diagnosticMessages.json]: This is supposed to be thrown only if all members are string literals. How about changing the wording to `A union of string literals cannot be used as an index signature parameter type directly. Use '[K in {0}]' instead.`?

File: tests/baselines/reference/indexerConstraints2.errors.txt (hunk 1 above)
Comment (kujon) [reply tests/baselines/reference/indexerConstraints2.errors.txt]: Given this suggestion https://github.com/Microsoft/TypeScript/pull/20726#discussion_r157842801, is it still applicable?

**Pull request merged**